ipn/ipnlocal/serve.go

```diff
@@ -257,6 +257,96 @@ func (b *LocalBackend) ServeConfig() ipn.ServeConfigView {
 	return b.serveConfig
 }
 
+// StreamFunnel turns on Funnel for the given HostPort, and keeps it on
+// for the duration that the context is open.
+//
+// When the context is done, Funnel is turned off for the port.
+//
+// During the stream, any incoming connections to the port are reported
+// out to the provided io.Writer, in the format ipn.FunnelRequestLog.
+func (b *LocalBackend) StreamFunnel(ctx context.Context, w io.Writer, hp ipn.HostPort) (err error) {
+	port, err := hp.Port()
+	if err != nil {
+		return err
+	}
+
+	// Turn on Funnel for the given HostPort.
+	sc := b.ServeConfig().AsStruct()
+	mak.Set(&sc.AllowFunnel, hp, true)
+	if err := b.SetServeConfig(sc); err != nil {
+		return err
+	}
+
+	// Defer turning off Funnel once stream ends.
+	cleanup := func() {
+		sc := b.ServeConfig().AsStruct()
+		delete(sc.AllowFunnel, hp)
+		err = errors.Join(err, b.SetServeConfig(sc))
+	}
+	defer cleanup()
+
+	f, ok := w.(http.Flusher)
+	if !ok {
+		return errors.New("writer not a flusher")
+	}
+
+	var writeErrs []error
+	writeToStream := func(log ipn.FunnelRequestLog) {
+		jsonLog, err := json.Marshal(log)
+		if err != nil {
+			writeErrs = append(writeErrs, err)
+			return
+		}
+		if _, err := fmt.Fprintf(w, "%s\n", jsonLog); err != nil {
+			writeErrs = append(writeErrs, err)
+			return
+		}
+		f.Flush()
+	}
+
+	// Hook up connections stream.
+	b.mu.Lock()
+	mak.Set(&b.funnelStreamers, port, &writeToStream)
+	b.mu.Unlock()
+
+	select {
+	case <-ctx.Done():
+	}
+
+	b.mu.Lock()
+	delete(b.funnelStreamers, port)
+	b.mu.Unlock()
+
+	return errors.Join(writeErrs...)
+}
+
+// maybeLogFunnelConnection logs out an ipn.FunnelRequestLog for the
+// connection if a streamer is currently active for the given destPort.
+func (b *LocalBackend) maybeLogFunnelConnection(destPort uint16, srcAddr netip.AddrPort) {
+	b.mu.Lock()
+	stream := b.funnelStreamers[destPort]
+	b.mu.Unlock()
```

Comment:
Should we use an RWMutex for the funnelStreamers?

Reply:
I'm using the general `sync.Mutex` from the LocalBackend - you suggesting one for particularly the streamers? Also interested in your thoughts on the new typing, which now supports multiple streamers watching one hostport at the same time.

```
funnelStreamers    map[uint16]map[uint32]func(ipn.FunnelRequestLog) // serve port => map of stream loggers (key is UUID)
```